### PR TITLE
Add support for multiarch builds

### PR DESCRIPTION
**Pull request by schmidp**
Hello,

we are running a K8S cluster with mostly ARM nodes. It would be great if `single-page-application-server` could support arm.

I have not verified the images yet as I have a problem with getting GitHub actually pushing the images to docker hub:

```
#33 [auth] library/1-nginx-stable-alpine:pull,push token for registry-1.docker.io
#33 DONE 0.0s

#28 exporting to image
#28 pushing layers 0.4s done
#28 ERROR: failed to push 1-nginx-stable-alpine: server message: insufficient_scope: authorization failed
```

I currently don't understand where the `library` comes from, any idea?

### CODE REVIEW

**Comment by PSanetra** [comment]
@schmidp Thank you for the PR! I will take a look at this soon. Supporting ARM makes definitely sense!

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -39,12 +39,22 @@ jobs:
...
-          password: ${{ secrets.DOCKER_PASSWORD }}
-          repository: ${{ secrets.DOCKER_REPOSITORY }}
-          tags: ${{ steps.target_tags.outputs.tags }}
-          always_pull: true
-          build_args: NGINX_TAG=${{ steps.target_nginx_tag.outputs.tag }}
+          context: .
+          platforms: linux/amd64,linux/arm64
+          push: true
+          pull: true
+          tags: ${{ steps.target_image_name.outputs.tags }}
```
Comment (PSanetra) [inline .github/workflows/main.yaml]: I think this was a breaking change with `build-push-action@v4`. I guess you need to prefix this tag with the full repository name (`${{ secrets.DOCKER_REPOSITORY }}`). See [these docs](https://docs.docker.com/build/ci/github-actions/local-registry/#:~:text=true%0A%20%20%20%20%20%20%20%20%20%20tags%3A-,localhost%3A5000/name/app%3Alatest,-%2D%0A%20%20%20%20%20%20%20%20name%3A).

File: .github/workflows/main.yaml [hunk 2]
```diff
@@ -39,12 +39,22 @@ jobs:
...
         if: ${{ steps.target_tags.outputs.tags != '' }}
         with:
-          username: ${{ secrets.DOCKER_USERNAME }}
-          password: ${{ secrets.DOCKER_PASSWORD }}
-          repository: ${{ secrets.DOCKER_REPOSITORY }}
-          tags: ${{ steps.target_tags.outputs.tags }}
-          always_pull: true
-          build_args: NGINX_TAG=${{ steps.target_nginx_tag.outputs.tag }}
+          context: .
+          platforms: linux/amd64,linux/arm64
```
Comment (PSanetra) [inline .github/workflows/main.yaml]: Can we support `linux/arm/v7`, too?

Comment (schmidp) [reply .github/workflows/main.yaml]: Screenshot 2023-08-01 at 10 29 32

I believe so, since the upstream nginx image supports it. I will add it.

**Comment by schmidp** [comment]
So I was now able to build and push images to https://hub.docker.com/r/openresearch/single-page-application-server/tags, but I have a strange problem.

When I pull the image on an ARM (MacBook Pro M1) machine:

```
docker pull --platform linux/arm64  openresearch/single-page-application-server:latest  
```

and then run it I get:

```
docker run -it openresearch/single-page-application-server
PROT_EXEC|PROT_WRITE failed.
```

The architecture of the files in the image are arm64 as I can do:

```
docker run -it --entrypoint '/bin/sh' openresearch/single-page-application-server                                                                           (master)single-page-application-server
/app $ 


docker run -it --entrypoint '/usr/sbin/nginx' openresearch/single-page-application-server                                                                      (master)single-page-application-server
2023/08/07 16:08:34 [notice] 1#1: using the "epoll" event method
2023/08/07 16:08:34 [notice] 1#1: nginx/1.24.0
2023/08/07 16:08:34 [notice] 1#1: built by gcc 12.2.1 20220924 (Alpine 12.2.1_git20220924-r10) 
2023/08/07 16:08:34 [notice] 1#1: OS: Linux 5.15.49-linuxkit-pr
2023/08/07 16:08:34 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2023/08/07 16:08:34 [notice] 1#7: signal 1 (SIGHUP) received, reconfiguring
```

But if I run:

```
docker run -it --entrypoint '/docker-entrypoint.sh' openresearch/single-page-application-server                                                                (master)single-page-application-server
PROT_EXEC|PROT_WRITE failed
```

I get the PROT_EXEC error again. Any idea what that could be?

UPDATE 7.8 20:03: so it's the `gomplate` command. The `-slim` version uses `upx` which apprently has problems on arm:

```
/usr/local/bin # wget https://github.com/hairyhenderson/gomplate/releases/download/v3.11.5/gomplate_linux-arm64-slim
Connecting to github.com (140.82.121.3:443)
Connecting to objects.githubusercontent.com (185.199.109.133:443)
saving to 'gomplate_linux-arm64-slim'
gomplate_linux-arm64 100% |*****************************************************************************************| 7702k  0:00:00 ETA
'gomplate_linux-arm64-slim' saved
/usr/local/bin # chmod +x gomplate_linux-arm64-slim 
/usr/local/bin # ./gomplate_linux-arm64-slim 
Illegal instruction
/usr/local/bin # wget https://github.com/hairyhenderson/gomplate/releases/download/v3.11.5/gomplate_linux-arm64
Connecting to github.com (140.82.121.3:443)
Connecting to objects.githubusercontent.com (185.199.109.133:443)
saving to 'gomplate_linux-arm64'
gomplate_linux-arm64 100% |*****************************************************************************************| 44.1M  0:00:00 ETA
'gomplate_linux-arm64' saved
/usr/local/bin # chmod +x gomplate_linux-arm64
/usr/local/bin # ./gomplate_linux-arm64 --version
gomplate version 3.11.5
```

**Comment by schmidp** [comment]
Ok, the ARM builds work and I *think* the PR could be merged. What I don't really understand is the interaction between the update and main workflow.

https://hub.docker.com/r/openresearch/single-page-application-server/tags

I am not sure how to get latest to be the same version as from the ones being pushed by pushing to main, but I don't think I changed anything in that regard.

**Comment by schmidp** [comment]
@PSanetra I used this now for close to a week on our arm cluster and it seems to work well.

**Comment by PSanetra** [comment]
@schmidp Thank you for your effort! I am currently on vacation. I will take another look when I am back 🙂

**Comment by PSanetra** [comment]
I have merged this PR manually to squash the commits and edit the commit message. Thank you! :slightly_smiling_face:

**Pull request closed without merging**